### PR TITLE
feat(core): optimise memory usage to reduce OS kills

### PR DESCRIPTION
**Pull request by ideoma**
- release O3 column buffers when TableWriter goes back to Writer Pool
- for WAL tables O3 column memory buffers are to be rounded to OS page size rather than 16Mb to granular track the used memory.
- on startup set the limit of writing memory to be the physical memory detected + swap - 0.5GiB. The assumption is that attempting to allocate beyond this limit will kill the process and it's better to fail the allocation with the exception sooner.
- track RSS memory allocated using malloc/realloc as a separate counter. Print RSS memory usage in TableWriter O3 processing.

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 42 / 48 (87.50%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/log/LogFactory.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/std/str/CharSink.java|5|8|62.50%|
|:large_blue_circle:|io/questdb/std/Unsafe.java|6|8|75.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalEventWriter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/log/AbstractLogRecord.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolColumnIndexer.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/pool/WriterPool.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/std/MemoryTag.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|16|16|100.00%|
|:large_blue_circle:|io/questdb/log/NullLogRecord.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 42 / 48 (87.50%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/log/LogFactory.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/std/str/CharSink.java|5|8|62.50%|
|:large_blue_circle:|io/questdb/std/Unsafe.java|6|8|75.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalEventWriter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/log/AbstractLogRecord.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolColumnIndexer.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/pool/WriterPool.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/std/MemoryTag.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|16|16|100.00%|
|:large_blue_circle:|io/questdb/log/NullLogRecord.java|1|1|100.00%|

File: core/src/main/java/io/questdb/cairo/pool/WriterPool.java [hunk 1]
```diff
@@ -131,7 +131,9 @@ public Map<CharSequence, Entry> entries() {
      * @return cached TableWriter instance.
      */
     public TableWriter get(TableToken tableToken, @NotNull String lockReason) {
-        return getWriterEntry(tableToken, lockReason, null);
+        TableWriter w = getWriterEntry(tableToken, lockReason, null);
```
[the remark this answers is not in the document]
Comment (ideoma) [reply core/src/main/java/io/questdb/cairo/pool/WriterPool.java:134]: I tried this test 

```Java
    @Test
    public void testFrequentCommitsPooled() throws Exception {
        assertMemoryLeak(() -> {
            int tableCount = 10;
            for(int i = 0; i < tableCount; i++) {
                ddl("create table abc" + i + " (" +
                        "i int," +
                        "l long," +
                        "d double," +
                        "c char," +
                        "s string," +
                        "sym symbol," +
                        "ts timestamp) timestamp(ts) partition by day BYPASS WAL;");
            }

            int rows = 100;
            Rnd rnd = new Rnd();
            for(int i = 0; i < rows; i++) {
                for(int t = 0; t < tableCount; t++) {
                    try (TableWriter writer = getWriter("abc" + t)) {
                        TableWriter.Row row = writer.newRow(rnd.nextLong(Timestamps.DAY_MICROS));
                        row.putInt(0, 1);
                        row.append();
                        writer.commit();
                    }
                }
            }
        });
    }
```

It creates 10 tables and does 100 transactions in each getting and putting back to the pool after each transaction. There is no measurable difference in the test before or after the change, it runs around 2.4 secs

**Review by bluestreak01 — APPROVED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 42 / 48 (87.50%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/log/LogFactory.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/std/str/CharSink.java|5|8|62.50%|
|:large_blue_circle:|io/questdb/std/Unsafe.java|6|8|75.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalEventWriter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/log/AbstractLogRecord.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolColumnIndexer.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/pool/WriterPool.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/std/MemoryTag.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|16|16|100.00%|
|:large_blue_circle:|io/questdb/log/NullLogRecord.java|1|1|100.00%|

**Pull request merged**